ballerina/msg_listener.bal

```diff
@@ -0,0 +1,246 @@
+// Copyright (c) 2025 WSO2 LLC. (http://www.wso2.org).
+//
+// WSO2 LLC. licenses this file to you under the Apache License,
+// Version 2.0 (the "License"); you may not use this file except
+// in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import ballerina/lang.runtime;
+import ballerina/log;
+import ballerina/task;
+
+# Represents the default error type.
+public type Error distinct error;
+
+# Represents the message store listener configuration,
+public type StoreListenerConfiguration record {|
+    # The interval in seconds at which the listener polls for new messages
+    decimal pollingInterval = 1;
+    # The maximum number of retries for processing a message. 
+    # If set to 0, the message will not be retried
+    int maxRetries = 3;
+    # The interval in seconds between retries for processing a message
+    decimal retryInterval = 1;
+    # If true, the message will be acknowledged with failure after the maximum number of retries is reached
+    boolean ackWithFailureAfterMaxRetries = false;
+    # An optional message store to store messages that could not be processed after the maximum 
+    # number of retries. When set, `ackWithFailureAfterMaxRetries` will be ignored
+    Store deadLetterStore?;
+|};
+
+# Represents a message store listener that polls messages from a message store and processes them.
+public isolated class StoreListener {
+
+    private Store messageStore;
+    private StoreService? messageStoreService = ();
+    private task:JobId? pollJobId = ();
+    private final StoreListenerConfiguration config;
+
+    # Initializes a new instance of Message Store Listener.
+    #
+    # + messageStore - The message store to retrieve messages from
+    # + config - The configuration for the message store listener
+    # + return - An error if the listener could not be initialized, or `()`
+    public isolated function init(Store messageStore, *StoreListenerConfiguration config) returns Error? {
+        self.messageStore = messageStore;
+        if config.maxRetries < 0 {
```

Comment:
Can't we use constraint module for this

Reply:
We could use that. But still we need to call validate to get the runtime validation. Since we don't want to have a constraint dependency just to have these simple validations, we have proceeded with this approach